core/modules/loader/DoList.h

```diff
@@ -0,0 +1,195 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2018 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ *
+ */
+#ifndef LSST_QSERV_LOADER_DOLIST_H_
+#define LSST_QSERV_LOADER_DOLIST_H_
+
+// system headers
+#include <chrono>
+#include <list>
+
+// Qserv headers
+#include "util/ThreadPool.h"
+
+namespace lsst {
+namespace qserv {
+namespace loader {
+
+
+class Central;
+
+
+class TimeOut {
+public:
+    using TimePoint = std::chrono::system_clock::time_point;
+    using Clock = std::chrono::system_clock;
+
+    TimeOut(std::chrono::milliseconds timeOut) : _timeOut(timeOut) {}
+
+    bool due() { return due(Clock::now()); }
+    bool due(TimePoint now) {
+        auto triggerDiff = std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastTrigger);
+        return (triggerDiff > _timeOut);
+    }
+
+    void triggered() { return triggered(Clock::now()); }
+    void triggered(TimePoint now) {
+        _lastTrigger = now;
+    }
+
+    std::chrono::milliseconds timeLeft(TimePoint now) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(now - _lastTrigger);
+    }
+
+    void setTimeOut(std::chrono::milliseconds timeOut) { _timeOut = timeOut; }
+    std::chrono::milliseconds getTimeOut() const { return _timeOut; }
+private:
+    // How much time since lastTrigger needs to pass before triggering.
+    std::chrono::milliseconds _timeOut; // default, 15 minutes
+    TimePoint _lastTrigger{std::chrono::seconds(0)};
+};
+
+
+/// Children of this class must be created with shared pointers.
+class DoListItem : public std::enable_shared_from_this<DoListItem> {
+public:
+    using Ptr = std::shared_ptr<DoListItem>;
+
+    DoListItem() = default;
```

Comment:
Igor recommended to me a pattern that I found to be useful when using `enable_shared_from_this`: make the constructors private and write a static factory function `makeNewDoListItem()` that returns a populated `DoListItem::Ptr`. This prevents anyone from creating a `DoListItem` that is not owned by a `shared_ptr`, and since calling `shared_from_this` on a non-pointer-owned-object leads to a misleading error ("bad weak_ptr" if I remember correctly), this prevents you from having to debug that issue & captures it at compile time.

Reply:
I have used that technique and I like it. I think making the constructor private causes problems with child classes, so forcing the use of a factory function becomes tricky. Maybe protected would work, but the child classes would also need factories.